### PR TITLE
Publish plugin to the Gradle Plugins Portal

### LINKED ISSUE #36: Plugin not found when using the new Plugin DSL notation in build.gradle
Plugin configuration:

*settings.gradle*:
```
pluginManagement {
  repositories {
      gradlePluginPortal()
      maven { url = uri("https://jitpack.io") }
  }
}
```
*build.gradle*:
```
plugins {
    [...]
    id 'com.epages.restdocs-openapi' version '0.2.1'
}
```

Generated urls are different for classic configuration using `buildscript` and a new one with `plugins`:
```
https://jitpack.io/com/github/epages-de/restdocs-openapi/restdocs-openapi-gradle-plugin/0.2.1/restdocs-openapi-gradle-plugin-0.2.1.pom
https://jitpack.io/com/epages/restdocs-openapi/com.epages.restdocs-openapi.gradle.plugin/0.2.1/com.epages.restdocs-openapi.gradle.plugin-0.2.1.pom
```
The first one is from `buildscript` and works.

### PR DESCRIPTION
**Pull request by chadlwilson**
**Motivation**

Use of the plugins DSL is much more concise and seems to be the way forward for Gradle. With the removal of need for Jitpack in #77 this appears to be easier now.

Resolves #36.

**Changes made**
* Followed guide [here](https://guides.gradle.org/publishing-plugins-to-gradle-plugin-portal/) and [here](https://plugins.gradle.org/docs/publish-plugin) to add the requisite metadata into the plugin.
* Included `mavenCoordinates` as my reading indicates we might need this since the plugin is already being published to bintray
* Added an additional deployment into the Travis build
* Updated the README to indicate how to use either approach

**For team to do**
* Register with https://plugins.gradle.org/user/register
* Get API keys into Travis
* If you could possibly publish an RC build (e.g. `0.9.1-rc1`) I can help test it. I initially updated the samples too; but probably better to have a build published first before we do that.

### CODE REVIEW

**Comment by mduesterhoeft** [comment]
Thanks for the PR. I will have a closer look tonight. It is really much better to go for the `plugins` block. This was just something that got a little bit out of sight.

**Comment by mduesterhoeft** [comment]
I have same issues publishing the plugin - I did not find a hint on how to solve this. Please see https://discuss.gradle.org/t/problems-publishing-an-existing-gradle-plugin-to-the-plugin-portal/31012?u=mduesterhoeft

**Comment by chadlwilson** [comment]
Ugh, yeah, that's a pain. I wonder if it's related to the `mavenCoordinates` section. Not sure how to address that bit.

**Comment by mduesterhoeft** [comment]
I also tried without the `mavenCoordinates` but the error persists

**Comment by chadlwilson** [comment]
Is there any way you think we could unblock this?

If we can't get ownership of `com.epages`, how about using `de.epages`?

**Comment by chadlwilson** [comment]
@mduesterhoeft Any chance of having another go at this?

There are some docs at https://plugins.gradle.org/docs/reclaiming for how to "claim" plugin/account and instructions of how to contact the admins via https://plugins.gradle.org/docs/get-help. Although this situation seems a little different, it does seem the process is to go via their Contact Us page?

**Comment by UgmaDevelopment** [comment]
Thanks for all the hard work on this issue. Was there ever any word back?

**Comment by ozscheyge** [comment]
Not that I'm aware of @UgmaDevelopment , this seems to be stuck.

**Comment by chadlwilson** [comment]
Can it perhaps just be published to a different location somehow as I suggested above (within Gradle Plugins portal) so we can move forward with current Gradle approaches to consuming plugins? :-/

**Comment by ozscheyge** [comment]
I'll have a look and will try to resolve this @chadlwilson

**Comment by ozscheyge** [comment]
```
> Task :restdocs-api-spec-gradle-plugin:publishPlugins FAILED
Could not reuse POM from pluginMaven publication because mavenCoordinates() was used
Publishing plugin com.epages.restdocs-api-spec version 0.9.9-SNAPSHOT

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':restdocs-api-spec-gradle-plugin:publishPlugins'.
> Failed to post to server.
  Server responded with:
  The user 'epages-dev' does not have write permissions to plugin [com.epages.restdocs-api-spec]
```
Trying to claim now.

File: restdocs-api-spec-gradle-plugin/build.gradle.kts
Comment (ozscheyge) [inline restdocs-api-spec-gradle-plugin/build.gradle.kts:11]: Needs to be version 0.12.0 now or this error will occur:

```
> Task :restdocs-api-spec-gradle-plugin:publishPlugins FAILED
Could not reuse POM from pluginMaven publication because mavenCoordinates() was used
Publishing plugin com.epages.restdocs-api-spec version 0.9.9-SNAPSHOT

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':restdocs-api-spec-gradle-plugin:publishPlugins'.
> Failed to post to server.
  Server responded with:
  This version of the com.gradle.plugin-publish plugin is no longer supported due to a critical security vulnerability.
  Please update to the latest version of the com.gradle.plugin-publish plugin found here:
     https://plugins.gradle.org/plugin/com.gradle.plugin-publish
  You can read more about this vulnerability here:
     https://blog.gradle.org/plugin-portal-update
```

**Comment by ozscheyge** [comment]
Seems like https://github.com/gradle/plugin-portal-requests/issues?q=is%3Aopen+is%3Aissue is the current way to claim groupIds. I'll prepare a request

File: restdocs-api-spec-gradle-plugin/build.gradle.kts
Comment (chadlwilson) [reply restdocs-api-spec-gradle-plugin/build.gradle.kts]: Ack, changed.

**Review by ozscheyge — APPROVED** [review]

**Comment by ozscheyge** [comment]
https://github.com/gradle/plugin-portal-requests/issues/73 worked like a charm:

```
> Task :restdocs-api-spec-gradle-plugin:publishPlugins
Could not reuse POM from pluginMaven publication because mavenCoordinates() was used
Publishing plugin com.epages.restdocs-api-spec version 0.9.9-SNAPSHOT
Thank you. Your new plugin com.epages.restdocs-api-spec has been submitted for approval by Gradle engineers. The request should be processed within the next few days, at which point you will be contacted via email.
Publishing artifact build/libs/restdocs-api-spec-gradle-plugin-0.9.9-SNAPSHOT.jar
Publishing artifact build/libs/restdocs-api-spec-gradle-plugin-0.9.9-SNAPSHOT-sources.jar
Publishing artifact build/libs/restdocs-api-spec-gradle-plugin-0.9.9-SNAPSHOT-javadoc.jar
Publishing artifact build/publish-generated-resources/pom.xml
Activating plugin com.epages.restdocs-api-spec version 0.9.9-SNAPSHOT

BUILD SUCCESSFUL in 2m 7s
25 actionable tasks: 2 executed, 23 up-to-date
```

Now waiting for approval:

Screenshot from 2021-01-15 15-09-23

**Comment by chadlwilson** [comment]
Great news @ozscheyge ! Thanks for following up on this :-)

**Comment by ozscheyge** [comment]
The gradle plugin has been approved:
```
plugins {
  id "com.epages.restdocs-api-spec" version "0.9.9-SNAPSHOT"
}
```
So far, that's only the version I published from my machine. I'd merge this, configure Travis and check the other follow-ups @chadlwilson mentioned.

**Comment by ozscheyge** [comment]
For future reference:

Adding secrets to travis config ( https://docs.travis-ci.com/user/encryption-keys/ ):

```
gem install travis
# Inside the repo directory
travis encrypt SOMEVAR="secretvalue"
```

Mapping environment variables to gradle.properties ( https://docs.gradle.org/current/userguide/build_environment.html#sec:project_properties ), so the follow two env vars need to be encrypted using the above method:
```
GRADLE_PUBLISH_KEY
GRADLE_PUBLISH_SECRET
```

**Comment by ozscheyge** [comment]
https://plugins.gradle.org/plugin/com.epages.restdocs-api-spec

Eventually it worked, `0.10.3` is currently first and latest version published to the Gradle plugins portal.

**Comment by chadlwilson** [comment]
Thanks @ozscheyge !

New published plugin seems to be working fine - was able to remove our `settings.gradle` hack.

```groovy
// Workaround until restdocs-api-spec is published to Gradle Plugins Portal per https://github.com/ePages-de/restdocs-api-spec/issues/36
pluginManagement {
    resolutionStrategy {
        eachPlugin {
            if (requested.id.id == "com.epages.restdocs-api-spec") {
                useModule("com.epages:restdocs-api-spec-gradle-plugin:${requested.version}")
            }
        }
    }
}
```

**Pull request merged**
